src/index.ts

```diff
@@ -1568,44 +1565,81 @@ export class BigQuery extends common.Service {
    */
   query(
     query: string | Query,
-    optionsOrCallback?:
-      | QueryOptions
-      | SimpleQueryRowsCallback
-      | QueryRowsCallback,
-    cb?: SimpleQueryRowsCallback | QueryRowsCallback
-  ): void | Promise<SimpleQueryRowsResponse> | Promise<QueryRowsResponse> {
-    let options =
+    optionsOrCallback?: QueryOptions | QueryRowsCallback,
+    cb?: QueryRowsCallback
+  ): void | Promise<QueryRowsResponse> {
+    const options =
       typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
     const callback =
       typeof optionsOrCallback === 'function' ? optionsOrCallback : cb;
-    this.createQueryJob(query, (err, job, resp) => {
-      if (err) {
-        (callback as SimpleQueryRowsCallback)(err, null, resp);
-        return;
-      }
-      if (typeof query === 'object' && query.dryRun) {
-        (callback as SimpleQueryRowsCallback)(null, [], resp);
-        return;
-      }
-      // The Job is important for the `queryAsStream_` method, so a new query
-      // isn't created each time results are polled for.
-      options = extend({job}, options);
-      job!.getQueryResults(options, callback as QueryRowsCallback);
-    });
+
+    query = typeof query === 'string' ? {query} : query;
+
+    if (query.table) {
+      query.table.getRows(options, callback!);
+      return;
+    }
+
+    this.runQuery_(query, options).then(
+      ([rows, nextQuery, resp]) => callback!(null, rows, nextQuery, resp),
+      callback!
+    );
   }
 
   /**
-   * This method will be called by `createQueryStream()`. It is required to
-   * properly set the `autoPaginate` option value.
+   * It is faster to use the `listTableData` rpc instead of `getQueryResults`,
+   * so when we run a query, we perform the following steps
+   *
+   *   1. create the query job
+   *   2. poll `getQueryResults` with `maxResults` set to 0 to check for job
+   *      completion
+   *   3. if we have a positive row count, call `listTableData` on the
+   *      destination table created by query job
    *
    * @private
    */
-  queryAsStream_(query: Query, callback?: SimpleQueryRowsCallback) {
-    if (query.job) {
-      query.job.getQueryResults(query, callback as QueryRowsCallback);
-      return;
+  async runQuery_(
+    query: Query,
+    options: QueryOptions
+  ): Promise<QueryRowsResponse> {
+    const [job, resp] = await this.createQueryJob(query);
+
+    if (query.dryRun) {
+      return [[], null, resp];
+    }
+
+    const queryOptions = Object.assign({}, options, {
+      autoPaginate: false,
+      maxResults: 0,
+    });
+
+    let rows, nextQuery, results;
+
+    do {
+      [rows, nextQuery, results] = await job.getQueryResults(queryOptions);
+    } while (!results!.jobComplete);
+
+    if (!Number(results!.totalRows)) {
```

Comment:
What does this check? Maybe there's a less exciting! way to write it?

Reply:
Gives back the number of rows inserted into the destination table. We need this for working with the Model API since models are created via SQL queries and will never return rows. They still create a destination table and querying against it results in an error.